### ISSUE TITLE
Add Publisher and Date Created to File Attachments

### ISSUE THREAD

**Issue by flittle8** [BCLibCoop/nnels#136]
If this issue is already being tracked in Github, let me know and I'll delete this.

Move Publisher and DC Date Created fields to each File Attachment. 
The Original Publisher, and DC Date Copyrighted and DC Date Issued, can remain in the Parent Node.

I'm not sure what the difference is between DC Date Copyrighted and DC Date Issued - which MARC fields are they taken from...?

**Comment by davekisly** [comment]
Ok, per discussion, i'll develop a VBO action script to grab those values from the parent node and populate each of the respective file attachment fields. This will work similar to other views where you click on selected records (or all) based on various filters (e.g., Publisher = null, Date = null).

**Comment by davekisly** [comment]
@flittle8 - can you review this View and let me know if this puts us on the right track for reviewing records? 

https://dev.nnels.ca/admin/content/cals/repo-items/attachments/publishers-and-dates

With respect to our convo re. date types yesterday, it does seem like the majority of records are assigned a date issued, as you can see in [this result set with "Issued" selected](https://dev.nnels.ca/admin/content/cals/repo-items/attachments/publishers-and-dates?title=&field_dc_date_value_1_op=not+empty&field_dc_date_value_1=&field_qualifier_date_value_op=and&field_qualifier_date_value%5B%5D=Issued&field_date_value_op=%3D&field_date_value%5Bvalue%5D=&field_date_value%5Bmin%5D=&field_date_value%5Bmax%5D=&field_date_created_value_op=%3D&field_date_created_value%5Bvalue%5D%5Bmonth%5D=&field_date_created_value%5Bvalue%5D%5Bday%5D=&field_date_created_value%5Bvalue%5D%5Byear%5D=&field_date_created_value%5Bmin%5D%5Bmonth%5D=&field_date_created_value%5Bmin%5D%5Bday%5D=&field_date_created_value%5Bmin%5D%5Byear%5D=&field_date_created_value%5Bmax%5D%5Bmonth%5D=&field_date_created_value%5Bmax%5D%5Bday%5D=&field_date_created_value%5Bmax%5D%5Byear%5D=&field_publisher_value_op=%3D&field_publisher_value=&field_publisher_source_value_op=%3D&field_publisher_source_value=&field_publisher_format_target_id_op=%3D&field_publisher_format_target_id%5Bvalue%5D=&field_publisher_format_target_id%5Bmin%5D=&field_publisher_format_target_id%5Bmax%5D=&field_record_set_nid=All&sort_by=nid&sort_order=DESC) vs. [this result set with Copyright selected](https://dev.nnels.ca/admin/content/cals/repo-items/attachments/publishers-and-dates?title=&field_dc_date_value_1_op=not+empty&field_dc_date_value_1=&field_qualifier_date_value_op=and&field_qualifier_date_value%5B%5D=Copyrighted&field_date_value_op=%3D&field_date_value%5Bvalue%5D=&field_date_value%5Bmin%5D=&field_date_value%5Bmax%5D=&field_date_created_value_op=%3D&field_date_created_value%5Bvalue%5D%5Bmonth%5D=&field_date_created_value%5Bvalue%5D%5Bday%5D=&field_date_created_value%5Bvalue%5D%5Byear%5D=&field_date_created_value%5Bmin%5D%5Bmonth%5D=&field_date_created_value%5Bmin%5D%5Bday%5D=&field_date_created_value%5Bmin%5D%5Byear%5D=&field_date_created_value%5Bmax%5D%5Bmonth%5D=&field_date_created_value%5Bmax%5D%5Bday%5D=&field_date_created_value%5Bmax%5D%5Byear%5D=&field_publisher_value_op=%3D&field_publisher_value=&field_publisher_source_value_op=%3D&field_publisher_source_value=&field_publisher_format_target_id_op=%3D&field_publisher_format_target_id%5Bvalue%5D=&field_publisher_format_target_id%5Bmin%5D=&field_publisher_format_target_id%5Bmax%5D=&field_record_set_nid=All&sort_by=nid&sort_order=DESC) from the above view (note the operators on the date qualifier and value fields).
You can also experiment with the Publishers field.
If this view works, then i can begin on the action script to copy over Publishers to the Format Publisher field and the DC Date (created) to Date Created?

**Comment by flittle8** [comment]
Thanks for creating this View; it looks like it will work to review the collection. Just wondering if, for clarity's sake, we could change the term "Format Publisher (File)" to "Publisher (File)"?
Also, not sure why there's an operation called "update file format"?
As for the Date Issued vs Date Copyrighted, I think we had discussed merging these(forget why?) but now I'm wondering if it's easiest to just leave them as is in the parent node? The most important thing is that Date Created is in the file attachment.

**Comment by flittle8** [comment]
As for the Date Issued vs Date Copyrighted - I think the original reason we had discussed merging these fields was to keep the record display simple for our users. So, I'm wondering if we could just display a field in the parent node called "Date". The value displayed to the user would not be prefixed by "issue" or "copyrighted". We could populate the value with whatever date is available in Drupal - either Date Issued or Date Copyrighted (if both exist, Date Copyrighted can take precedence). How does that sound?

**Comment by davekisly** [comment]
Quick reply:
- I've updated the view, renaming the Format Publisher label and removing the "update" option (which was actually just a placeholder until i add a new custom action).
- your suggestion re. consolidating Date Issued and Date Copyrighted sounds reasonable. I'll work on a script to coordinate that this week, as well as move the Date Created into the file attachment.

**Comment by davekisly** [comment]
hi @flittle8 - i've just uploaded / committed further changes to this operation: https://dev.nnels.ca/admin/content/cals/repo-items/attachments/publishers-and-dates

The view is a bit tricky, but essentially it returns all repo items with Publisher values OR with Date Created values. The operation you see attached to this should allow you to copy those values to the newly created fields on the File / Field Collection, namely Publisher (awkwardly labelled Format Publisher), Place of Publication and Date Created.

I did have some issues w/ dates, so please bang away on this and let me know if u see any anomalies. If not, i'll copy some of this to PROD.

As noted, for the time being, i think its safest to retain the original fields until i get the import parsers updated to directly populate the field collections when run.

**Comment by flittle8** [comment]
hi @davekisly - thanks for creating this view. 

I think the terminology is throwing me off in terms of how to use this, so I've attached a screenshot of how I think we should label things. Let me know if it makes sense to you, or not.

The operation seems to be working; however I think bundling the 2 operations into 1  (Copy Date Created & Copy Original Publisher To Field Collection) could be problematic. This is because Date Created _should_ be associated with the file(s) but the Original Publisher would not be associated with them. For example, node [99945](https://dev.nnels.ca/items/à-la-recherche-du-bout-du-monde-ressource-électronique): BC Libraries Coop published the DAISY but this operation copies over the original publisher (incorrect) along with the date created (correct). 

I think this operation works perfectly with those files that we publish directly; for example, if we publish an epub by Greystone Books and Greystone Books is also the Original Publisher. But it won't work with those files we publish ourselves as the Original Publisher will be different (but the Date Created would be correct...). 

![date_created_and_publisher_view2](https://user-images.githubusercontent.com/23460746/31143759-28cde0b4-a833-11e7-92ad-b6f706c86be9.png)

**Comment by davekisly** [comment]
hi - a couple quick replies back:
- the view should return items where there's a publisher value **or** there's a date-created value. So, from that perspective, bundling the two operations together might end up saving you time as you'll be covering one or both field migrations at the same time. But, i can unbundle them as well and in that case, i might suggest we stick them into different views as well (one for publishers, the other for dates)? Just confirm your preference, this wouldn't take long to set-up.
- i know the labeling is a bit icky, but to confirm, **Place of Publication** and **Format Publisher** are the values on the **file attachment** (field collection) that have been copied over from the node's **Publisher** field. When it comes to the actual display (both for admins and end-users), we can change those labels as needed.
- to confirm, however, is it the **Publisher** field i should be copying, or the **Source Publisher** - is it the latter that you are calling the "Original Publisher"? The original post in this issue thread simply notes "Move Publisher", so my assumption is you mean the Publisher field, i.e., the one i've labeled on the view as **Publisher (Node)**.

**Comment by davekisly** [comment]
hi @flittle8 - quick note to say i've just committed a couple updates to DEV, namely the two views we spoke about and the corresponding functions:
- https://dev.nnels.ca/admin/content/cals/repo-items/attachments/dates
- https://dev.nnels.ca/admin/content/cals/repo-items/attachments/publishers

Give them a test and let me know if this addresses what we chatted about?

**Comment by flittle8** [comment]
Hi Dave - the updates all look good, thanks. 

Here is the URL for the record where Date Created doesn't copy over:
https://dev.nnels.ca/items/started-early-took-my-dog 

Also, re: my prior comment "Original Publisher field on the parent node should draw its
values from the _Publisher (source)_ field", I was referring to the Original Publisher information that you see on the parent node: https://dev.nnels.ca/items/crossed-bones

**Comment by davekisly** [comment]
@flittle8 - thanks, i just tested that and you're right, the date isn't being copied. I'll review this later today.

**Comment by davekisly** [comment]
@flittle8 - just committed an update and re-parsed the "Started Early" title above: date now displays as 2013.

Along w/ some of the other display / labeling issues you mentioned in emails, we (someone) will also want to tidy up some of the formatting on the File Attachments views on the node detail pages: e.g., we'll want to align the publisher and date labels so they are consistent w/ the other fields. Maybe put these into a separate ticket?

**Comment by flittle8** [comment]
Hi Dave, just wondering if anything remains on this issue before we can move to PROD?

**Comment by davekisly** [comment]
hi Farrah, i actually did begin moving selected items over, e.g., the publisher and the date view, plus the fields on the attachment.

I did hold off on pulling the cals_importer module updates to Prod to give me some time to make sure the work we've done over the past couple months hasn't introduced any regression errors into code. I haven't seen anything egregious but perhaps i should double-check w/ @gitschatz to make sure none of his code would bleed thru to the regular import / parsing stuff?

Barring any issues, i can move that code up as early as this evening or tomorrow morning.

**Comment by gitschatz** [comment]
@davekisly I believe it to be free of side effects. But run a quick parse on dev to be sure.

My process is only triggered by the snippet button and the test function is tied to certain stub generation helpers. For example: _cals_importer_exporter_parse_controlfields() doesn't appear to be called anywhere in the parser stack.

**Comment by davekisly** [comment]
thanks @gitschatz. That's as i expected. Will begin moving stuff up this eve, then.

**Comment by flittle8** [comment]
Thanks. I noticed that some bits have been moved to Prod already. 
Could we change the name of the below field? If renaming fields is something I could do, let me know, then I won't have to bother you with this stuff :)

On the File Resource page, there's a field called "Publisher Format", let's rename this to "Name of Publisher".

**Comment by davekisly** [comment]
The field label changes are not difficult, but you might want to leave that with me for the time being (or hand off to Jonathan) ... esp. on PROD. I can walk you thru the process at some point, but its also good to know some of the other impacts this might or might not have (e.g., in various displays, etc.). In the meantime, i've just updated the label accordingly.

And, i've moved up the module itself so you will now see the same options on PROD that you have on DEV:
- https://nnels.ca/admin/content/cals/repo-items/attachments/dates
- https://nnels.ca/admin/content/cals/repo-items/attachments/publishers

Do a couple test runs with a couple records here at a time. Then, before you dive into larger runs, it might be useful to quickly run a database backup: https://nnels.ca/admin/content/cals/repo-items/attachments/publishers

I'm happy to touch base this eve or Thurs if you like.

**Comment by davekisly** [comment]
@gitschatz - had a brief scare when i moved up the cals_importer - site whitescreened. No biggie, quickly resolved by commenting out a couple lines in the .exporter.inc file:
- //require 'File/MARCXML.php';
- //require 'File/MARC/Lint.php';

So, i guess we'll need the the MARC_FILE installed on PROD at some point in the not too distant future.

**Comment by gitschatz** [comment]
Oh my bad, I thought we had the Pear packages installed. Adding them now.

Also add it as dependency in .info?

**Comment by davekisly** [comment]
Thanks!

Re. adding this as a dependency, i'm not certain this is possible (but could be overlooking options): https://www.drupal.org/docs/7/creating-custom-modules/writing-module-info-files-drupal-7x#dependencies

**Comment by flittle8** [comment]
Wondering if we can do a quick change to the user display - could we delete the "publisher" field from the parent node user display? We can display the "Source Publisher" field instead (under Source Information). Screenshot attached.
![user_display_node](https://user-images.githubusercontent.com/23460746/31971297-0d734284-b8d0-11e7-9dfc-7b37aed94341.png)

**Comment by davekisly** [comment]
Ok,  can do. Just to clarify, do you want to use "Source Publisher" as the label? Or change that to "Publisher"?

**Comment by gitschatz** [comment]
@davekisly huh, I guess it is a hook_libraries_info (Libraries API) and a dependency line in .info

Example: https://www.drupal.org/docs/7/modules/libraries-api-7x/using-libraries-api-2x-as-a-module-developer

**Comment by flittle8** [comment]
I was thinking we delete the entire Publisher field from the user display and display the Source Publisher field instead (since the data in that Publisher field is no longer applicable, Publisher is now associated with the individual files).

**Comment by davekisly** [comment]
@gitschatz - ok, maybe we should test that on DEV first?
@flittle8  - sure, i was just asking about the label itself, not the field value. I'll update shortly and we can go from there.

**Comment by flittle8** [comment]
@davekisly - Ah ok, looking at DEV, I recall we agreed on the label "Original Publisher".

**Comment by davekisly** [comment]
ok, have a look here: https://nnels.ca/items/nova-scotia-place-names-place-name-origins-attractions-legends-characters-history-and-firsts

This field looks ok, but is apparently being modified by some css injecting the subheading content ("Source Information") above the field itself. I'm not sure what role this css plays; I did try to bypass this w/ a custom field, etc., but that's less preferable so i'll leave the view/listing as is for now.

@gitschatz, do you have any ideas on this?

**Comment by flittle8** [comment]
Looks like I can't view that link until I'm finished processing a batch of records. Are you able to use PROD while I process batches?

**Comment by davekisly** [comment]
yep, just open a second browser window or tab. It shouldn't eat all the site resources (that's sort of the purpose of the batch api), but let me know if you can't do the above.

**Comment by flittle8** [comment]
Yes, would be great to hide that "Source information" subheading, as technically everything under the "Details" heading is the source information.

**Comment by flittle8** [comment]
Could we add the ability to search for keywords in the filepath string to both the update date and update publisher views? This would help with updating records and determining the values for date and publisher.

**Comment by davekisly** [comment]
hi - re. "Source information", i'd suggest we add that to a separate ticket (and probably assign to @gitschatz who can, in turn, assign to me if he prefers).

Re. filepath search, i'll see what i can do.

**Comment by davekisly** [comment]
Ok, the filepath search is now incorporated. I've set the default operator to "contains" (which allows partial string searches), but you might also experiment w/ "starts with" and also perhaps the "regular expression" option (e.g., pd|restricted would find items with either pd or restricted).

**Comment by flittle8** [comment]
@davekisly re opening a second browser or tab - yes, that lets me do other activities on the site while I process records, but all of the processing actions appear at the top of every page I navigate to, as in attached screenshot. Is there a way to hide this...? 
![can_see_current_operation](https://user-images.githubusercontent.com/23460746/32017573-fe503f62-b97b-11e7-8515-48c7282abad9.png)

**Comment by davekisly** [comment]
Those are messages i added to the script to provide links to the items processed (just in case u need to review on that page). Because of the way they are initialized (e.g., as $_SESSION variables), Drupal will add them to whatever page you land on. 

I can, however, easily suppress those if they're not of real value..

**Comment by flittle8** [comment]
OK thanks. Yes, I think it’s of value, but only on the page where the operations are occurring, so I can see any errors, how many nodes updated etc…  If it’s a matter of suppressing it from all pages or no pages, let’s keep it for now. If it gets really annoying and makes it hard to work, I’ll let you know. :)

**Comment by flittle8** [comment]
Hi - I just noticed there's a bug in the "Copy Dublin Core Date from Node to Field Collection" operation in [the date view](https://nnels.ca/admin/content/cals/repo-items/attachments/dates) . When I select "date copyrighted" as the date type I want to copy over, it copies over the timestamp instead of the date copyrighted.

**Comment by davekisly** [comment]
I just tested w/ the first record **[on this list](https://nnels.ca/admin/content/cals/repo-items/attachments/dates?title=&field_dc_date_value_1_op=%3D&field_dc_date_value_1=2016&field_qualifier_date_value_op=or&field_qualifier_date_value%5B%5D=Copyrighted&field_date_value_op=%3D&field_date_value%5Bvalue%5D=&field_date_value%5Bmin%5D=&field_date_value%5Bmax%5D=&field_date_created_value_op=%3D&field_date_created_value%5Bvalue%5D%5Bmonth%5D=&field_date_created_value%5Bvalue%5D%5Bday%5D=&field_date_created_value%5Bvalue%5D%5Byear%5D=&field_date_created_value%5Bmin%5D%5Bmonth%5D=&field_date_created_value%5Bmin%5D%5Bday%5D=&field_date_created_value%5Bmin%5D%5Byear%5D=&field_date_created_value%5Bmax%5D%5Bmonth%5D=&field_date_created_value%5Bmax%5D%5Bday%5D=&field_date_created_value%5Bmax%5D%5Byear%5D=&field_record_set_nid=All&status=1&field_publisher_value_op=contains&field_publisher_value=&field_file_format_value_op=or&field_file_format_value%5B%5D=5&field_s3_path_value_op=contains&field_s3_path_value=&sort_by=nid&sort_order=DESC)** <- this worked for me.

Can you provide me w/ specific examples so i can test?

**Comment by davekisly** [comment]
Ok, i tested the same records on PROD (the **[2 in this list](https://nnels.ca/admin/content/cals/repo-items/attachments/dates?title=&field_dc_date_value_1_op=%3D&field_dc_date_value_1=2016&field_qualifier_date_value_op=or&field_qualifier_date_value%5B0%5D=Copyrighted&field_date_value_op=%3D&field_date_value%5Bvalue%5D=&field_date_value%5Bmin%5D=&field_date_value%5Bmax%5D=&field_date_created_value_op=%3D&field_date_created_value%5Bvalue%5D%5Bmonth%5D=&field_date_created_value%5Bvalue%5D%5Bday%5D=&field_date_created_value%5Bvalue%5D%5Byear%5D=&field_date_created_value%5Bmin%5D%5Bmonth%5D=&field_date_created_value%5Bmin%5D%5Bday%5D=&field_date_created_value%5Bmin%5D%5Byear%5D=&field_date_created_value%5Bmax%5D%5Bmonth%5D=&field_date_created_value%5Bmax%5D%5Bday%5D=&field_date_created_value%5Bmax%5D%5Byear%5D=&field_record_set_nid=All&status=1&field_publisher_value_op=contains&field_publisher_value=&field_file_format_value_op=or&field_file_format_value%5B0%5D=5&field_s3_path_value_op=contains&field_s3_path_value=&sort_by=nid&sort_order=DESC)** and am seeing the copyright date being copied over when selected.

**Comment by flittle8** [comment]
Hi Dave - can you see this page <https://nnels.ca/admin/content/cals/repo-items/attachments/dates?title=&field_dc_date_value_1_op==&field_dc_date_value_1=&field_qualifier_date_value_op=or&field_date_value_op==&field_date_value%5Bvalue%5D=&field_date_value%5Bmin%5D=&field_date_value%5Bmax%5D=&field_date_created_value_op==&field_date_created_value%5Bvalue%5D%5Bmonth%5D=&field_date_created_value%5Bvalue%5D%5Bday%5D=&field_date_created_value%5Bvalue%5D%5Byear%5D=&field_date_created_value%5Bmin%5D%5Bmonth%5D=&field_date_created_value%5Bmin%5D%5Bday%5D=&field_date_created_value%5Bmin%5D%5Byear%5D=&field_date_created_value%5Bmax%5D%5Bmonth%5D=&field_date_created_value%5Bmax%5D%5Bday%5D=&field_date_created_value%5Bmax%5D%5Byear%5D=&field_record_set_nid=All&status=1&field_publisher_value_op=allwords&field_publisher_value=UBC%20Press&field_file_format_value_op=or&field_s3_path_value_op=contains&field_s3_path_value=NNELS&sort_by=nid&sort_order=DESC>? I tried with these 2 records and it copies the timestamp: 879888 and 879882

**Comment by davekisly** [comment]
ok, thanks good example. Interesting. Will investigate on DEV (and local).

**Comment by flittle8** [comment]
I started processing a big batch of records this morning (4952 files) using the Update Date view. After 4 hours of processing and when it was over 90% complete, it decided to stop with the attached error message. I'm not sure if this happened because the batch is too big, is there a limit to the # of files we can update at once?
![processing_error](https://user-images.githubusercontent.com/23460746/32023103-2d8e111c-b98d-11e7-9534-e8838cba5dfc.png)

**Comment by flittle8** [comment]
thanks!

**Comment by davekisly** [comment]
@flittle8 - will take a bit of time to review the Copyrighted issue above. In the meantime, please do proceed with smaller batches. For example, start w/ n=100 just so you can ensure things are working and also to minimize overhead. The batch process "spreads" out the overhead quite effectively, but its still an intensive process due to the number of entities that are updated (in the DB) - always a relatively "expensive" affair.

**Comment by davekisly** [comment]
ack, its an interesting issue. The underlying field values for the date qualifiers should = "Copyrighted", but for some reason, these records are stored as "Copyright". The parser thus skipped over those and defaulted to the timestamp. I've added a temp fix to map Copyright onto Copyrighted and so we're back in biz. 

Please test further!

**Comment by flittle8** [comment]
Hi Dave - the last issue of the day, promise. It's related to the Update publishers view. 
I'm trying to use the "Update Publisher/Place of Publication" operation; however I don't think it's working. For example, I want to batch update the publisher field (with new values) for the files on [this page](https://nnels.ca/admin/content/cals/repo-items/attachments/publishers?title=&field_publisher_value_op=not+empty&field_publisher_value=&field_publisher_source_value_op=%3D&field_publisher_source_value=&field_publisher_format_value_op=empty&field_publisher_format_value=&field_place_of_publication_value_op=%3D&field_place_of_publication_value=&status=1&field_record_set_nid=All&field_s3_path_value_op=contains&field_s3_path_value=RBdigital&sort_by=field_publisher_format_value&sort_order=ASC ). I select a few files,  I click "Update publisher/place of publication" and I enter the values for “name of publisher” and “place of publication”.  It says it processes successfully but I don’t see that any changes have actually been made to the records. 

(Also, strangely, after it processes, it returns me to the results page but the files that I have "updated" are mysteriously not on the page any longer nor do they appear when I search for them by title...)

Thanks for looking into this.

**Comment by flittle8** [comment]
Thank you!!

**Comment by davekisly** [comment]
@flittle8 - will review tomorrow, but one reason they may be disappearing is the search is set to "Format Publisher (file)" = NULL. So, when it returns you to the search page, the VBO may be filtering out the very titles you've processed. :)

**Comment by flittle8** [comment]
Ah, that was just the default which I didn’t notice, thanks. 
I just realized what is going on: when I click “update/publisher place of publication” it is updating the publisher (file) value and not the publisher (node). For some reason, I was expecting it to update the publisher (node). Then I was planning to copy over the new publisher(node) values to the field collections (using the Copy Publisher to Field Collection operation). 
So I think  the operation is actually working. :)

**Comment by flittle8** [comment]
Would we be able to add a "Copy Source Publisher to Field Collection" operation to the Update publisher view? I'm finding we have tons of records where the Source Publisher (node) needs to be copied to the field collection values (instead of the publisher node values being copied). As an example, all the books in this [results list ](https://nnels.ca/admin/content/cals/repo-items/attachments/publishers?title=&field_publisher_value_op=contains&field_publisher_value=cooperative&field_publisher_source_value_op=not+empty&field_publisher_source_value=&field_publisher_format_value_op=not+empty&field_publisher_format_value=&field_place_of_publication_value_op=%3D&field_place_of_publication_value=&status=1&field_record_set_nid=All&field_s3_path_value_op=contains&field_s3_path_value=CA%2FSK&field_s3_path_value_1_op=contains&field_s3_path_value_1=publisher&sort_by=field_publisher_format_value&sort_order=ASC) should have the Source Publisher(node) value in the Field Collection.

This operation would be super useful in cleaning up our records. Otherwise I have to update each record manually. The only issue is that values in the Publisher (node) and Source Publisher(node) will be messy or inaccurate, but at least the values attached to the files will be cleaned up and accurate.

**Comment by davekisly** [comment]
Shouldn't be a problem. Do you prefer this as a distinct "button" (operation), or implemented similar to what we've done with the Dates, i.e., with the second page giving you a select field where u indicate which of the fields you are copying from?

**Comment by flittle8** [comment]
Hi Dave, 
I think similar to how we’ve done it with Dates would work well. For example, we click “Copy Publisher to Field Collection” and then have the option to copy the Publisher (Node) or Source Publisher(Node) values.
Thank you, thank you.
Farrah

**Comment by davekisly** [comment]
Ok, i've got the basic process working on my local, but have a couple questions about the Source Publisher. The regular "Publisher" field is multi-valued because we often (usually?) parse both place and publisher into 2 entries. What about Source? Its multivalued in the way its set up, but in practice, do you add multiple entries?

I ask because the Publisher field is a bit tricky in the way we concatenate multiple values together with a ":" (colon). If we don't need all that with Source Publisher, i'm going to simplify what happens there.

**Comment by davekisly** [comment]
Actually, let me expand on that question. You available for a quick skype call, Farrah?

**Comment by davekisly** [comment]
hi @flittle8 - further to our discussion, i've just committed updates that should facilitate copying the **field_publisher_source** values to the file collection. As discussed, the **field_publisher** parsing is actually a bit complex due to how it teases apart the **Publisher** and the **Place of Publication** values. It can do this primarily because **field_publisher** is multi-valued and the values were added in a predictable order. 

**field_publisher_source** was never really set up this way and so for the time being, we've agreed to move the entire entry into the Field Collection publisher value (we won't hive off Place, etc.). 

At a later phase, you may decide to normalize the publisher values on the Field Collections ... but at this point, its a pretty blunt process. 

I've also added a 3rd S3 Path filter into the views.

Please test with a couple records first before doing larger batches.

**Comment by flittle8** [comment]
Perfect, thanks. 
Also, could we add another 2 S3 path filters to the views…? (so we’ll have 5 on each view now).

**Comment by davekisly** [comment]
Extra S3 path filters now in place.

**Comment by flittle8** [comment]
I think I've finished updating the publisher data for all files in the repo now. I'm wondering if we could now make these fields visible to the users? (but not the date field, yet). As in DEV, they should be called: "Published in: _YYYY_" and "Published by: _name of publisher, location_". Like you mentioned in our convo, could we also suppress these fields if there is no data in them. Thanks!

**Comment by davekisly** [comment]
Its pretty easy to add them into the various public views, but the bigger issue here is the positioning of the fields / labels. We probably need to tweak the CSS a bit to sync them flush w/ the other fields. I think that's better worked out on DEV, possibly something @gitschatz would want to do?

**Comment by flittle8** [comment]
OK, thanks, I've created a new ticket here #145

**Comment by flittle8** [comment]
In the Update dates view, setting Date Created (File) to IS EMPTY (NULL) doesn’t seem to work, it returns all records even those where date created file contains data. [Search results](https://nnels.ca/admin/content/cals/repo-items/attachments/dates?title=&field_dc_date_value_1_op=%3D&field_dc_date_value_1=&field_qualifier_date_value_op=or&field_date_value_op=%3D&field_date_value%5Bvalue%5D=&field_date_value%5Bmin%5D=&field_date_value%5Bmax%5D=&field_date_created_value_op=empty&field_date_created_value%5Bvalue%5D%5Bmonth%5D=&field_date_created_value%5Bvalue%5D%5Bday%5D=&field_date_created_value%5Bvalue%5D%5Byear%5D=&field_date_created_value%5Bmin%5D%5Bmonth%5D=&field_date_created_value%5Bmin%5D%5Bday%5D=&field_date_created_value%5Bmin%5D%5Byear%5D=&field_date_created_value%5Bmax%5D%5Bmonth%5D=&field_date_created_value%5Bmax%5D%5Bday%5D=&field_date_created_value%5Bmax%5D%5Byear%5D=&field_record_set_nid=All&status=1&field_publisher_value_op=contains&field_publisher_value=&field_file_format_value_op=or&item_id_op=regular_expression&item_id%5Bvalue%5D=&item_id%5Bmin%5D=&item_id%5Bmax%5D=&field_s3_path_value_op=contains&field_s3_path_value=&field_s3_path_value_1_op=not&field_s3_path_value_1=&field_s3_path_value_2_op=contains&field_s3_path_value_2=&field_s3_path_value_3_op=contains&field_s3_path_value_3=&field_s3_path_value_4_op=contains&field_s3_path_value_4=&sort_by=nid&sort_order=DESC).

**Comment by davekisly** [comment]
Yep, seems to be a bug specifically on "exposed" filters (ie., searchable fields). I can clone this view display w/ the created = NULL hard-coded into the query, and it should work, just you can't unselect that via the interface. Would that suffice for your immediate needs?

If i had more time, i'd try to troubleshoot the issue ... but its not necessarily a trivial / quick fix.

**Comment by flittle8** [comment]
I need to be able to toggle between empty, not empty, etc. so I’ll see if I can find some workaround for now, and I’ll let you know if I need you hard-code created=NULL into the search. Thanks!

**Comment by flittle8** [comment]
When there's a date in the node but it's set to 'none', this date isn't copied over during the copy date issued or copy date copyrighted operation. You can see in this record that in [this record](https://nnels.ca/items/fierce-beauty-choosing-stand-what-matters-most), the timestamp (2017) was copied over to the EPUB2, instead of the date in the node (211). 
If there's a date in the node that's set to 'none', let's assume this is a date issued.

**Comment by flittle8** [comment]
Hi Dave- looks like I do need this functionality, so would you be able able clone the view with date created(file)=NULL? Also, just want to check that I’ll still be able to use the other view (where date created is not null)?
Thanks

**Comment by davekisly** [comment]
hi Farrah - Here's a listing w/ date created (file) = NULLL: https://nnels.ca/admin/content/cals/repo-items/attachments/dates/null

Note that you can always open the same view display (with the same filters) in 2 separate tabs. Just submit on your first and then copy the resulting URL into a second tab. I mention this because this might be an easy way to track (via the second tab) the items you've worked on following your updates in the first tab.

On a separate note, i'll have a look this evening at the issue you describe above re. the date not being copied over where the Qualifier = NONE. It sounds as though "NONE" needs to be added to the list of options you can select from?

**Comment by davekisly** [comment]
@flittle8 - please have a look here and let me know if the new "No Dublin Core Qualifier" select option gives you what you need in terms of copying dates with null qualifiers? (Note also the DC **Date Value (node)** and **Date Type / Qualifer (node)** filters i set to return the result list in question.)

**Comment by flittle8** [comment]
Hi Dave, thanks for doing this, but it’s not quite what we need. We need all dates that don’t have a qualifier (i.e. it’s set to NONE) to be treated as if they have the qualifier ISSUED (so, the qualifier “none” = qualifier ”issued). I have no way to tell which records have a date qualifier of “none” at the moment without opening the record. We need to treat these dates as "date issued” so that they get copied over in that specific operation. Let me know if that makes sense? Thanks!

**Comment by davekisly** [comment]
I don't think i follow, so i'll try to explain what i've set up and you tell me what i need to fix. **[This URL](https://nnels.ca/admin/content/cals/repo-items/attachments/dates?title=&field_dc_date_value_1_op=not+empty&field_dc_date_value_1=&field_qualifier_date_value_op=empty&field_date_value_op=%3D&field_date_value%5Bvalue%5D=&field_date_value%5Bmin%5D=&field_date_value%5Bmax%5D=&field_date_created_value_op=%3D&field_date_created_value%5Bvalue%5D%5Bmonth%5D=&field_date_created_value%5Bvalue%5D%5Bday%5D=&field_date_created_value%5Bvalue%5D%5Byear%5D=&field_date_created_value%5Bmin%5D%5Bmonth%5D=&field_date_created_value%5Bmin%5D%5Bday%5D=&field_date_created_value%5Bmin%5D%5Byear%5D=&field_date_created_value%5Bmax%5D%5Bmonth%5D=&field_date_created_value%5Bmax%5D%5Bday%5D=&field_date_created_value%5Bmax%5D%5Byear%5D=&field_record_set_nid=All&status=1&field_publisher_value_op=contains&field_publisher_value=&field_file_format_value_op=or&item_id_op=regular_expression&item_id%5Bvalue%5D=&item_id%5Bmin%5D=&item_id%5Bmax%5D=&field_s3_path_value_op=contains&field_s3_path_value=&field_s3_path_value_1_op=not&field_s3_path_value_1=&field_s3_path_value_2_op=contains&field_s3_path_value_2=&field_s3_path_value_3_op=contains&field_s3_path_value_3=&field_s3_path_value_4_op=contains&field_s3_path_value_4=&sort_by=nid&sort_order=DESC)** returns records with at least one date entry (on the node) that include **no** date qualifier. 

For example, if you look at the first record (id = 891147), you will see the first date is unqualified (1988), the second date (2017) is qualified (Created). Currently, if you run the VBO and select "No Dublin Core Qualifier", it should copy 1988 over to the **Date Created** on the **File**. Is that not the action you need here?

Or, by "_treated as if they have the qualifier ISSUED_", do you mean the VBO should add the qualifier "issued" to those dates -on the **node**- that currently lack a qualifier?

**Comment by flittle8** [comment]
OK, I see what you’ve done, thanks. But yes, I was meaning that the qualifier “issued” should be attached to those dates that lack a qualifier. Then I can process them as if they are date issued. But let’s hold off on this for the moment, this view might do the trick...

**Comment by flittle8** [comment]
Completed updating dates and publishers, going to close this ticket.
